qiskit/qasm3/printer.py

```diff
@@ -15,81 +15,73 @@
-    Expression,
-    GtOperator,
-    Header,
-    IO,
-    IOModifier,
-    Identifier,
-    Include,
-    Integer,
-    LtOperator,
-    PhysicalQubitIdentifier,
-    Pragma,
-    Program,
-    ProgramBlock,
-    QuantumArgument,
-    QuantumBarrier,
-    QuantumDeclaration,
-    QuantumGateCall,
-    QuantumGateDefinition,
-    QuantumGateModifier,
-    QuantumGateModifierName,
-    QuantumGateSignature,
-    QuantumMeasurement,
-    QuantumMeasurementAssignment,
-    QuantumReset,
-    Range,
-    ReturnStatement,
-    SubroutineCall,
-    SubroutineDefinition,
-    SubscriptedIdentifier,
-    Version,
-)
+from . import ast
 
 
 class BasicPrinter:
     """A QASM 3 AST visitor which writes the tree out in text mode to a stream, where the only
     formatting is simple block indentation."""
 
     _CONSTANT_LOOKUP = {
-        Constant.pi: "pi",
-        Constant.euler: "euler",
-        Constant.tau: "tau",
+        ast.Constant.pi: "pi",
+        ast.Constant.euler: "euler",
+        ast.Constant.tau: "tau",
     }
 
     _MODIFIER_LOOKUP = {
-        QuantumGateModifierName.ctrl: "ctrl",
-        QuantumGateModifierName.negctrl: "negctrl",
-        QuantumGateModifierName.inv: "inv",
-        QuantumGateModifierName.pow: "pow",
+        ast.QuantumGateModifierName.ctrl: "ctrl",
+        ast.QuantumGateModifierName.negctrl: "negctrl",
+        ast.QuantumGateModifierName.inv: "inv",
+        ast.QuantumGateModifierName.pow: "pow",
+    }
+
+    _FLOAT_WIDTH_LOOKUP = {
+        ast.FloatType.half: "16",
+        ast.FloatType.single: "32",
+        ast.FloatType.double: "64",
+        ast.FloatType.quad: "128",
+        ast.FloatType.oct: "256",
     }
 
     # The visitor names include the class names, so they mix snake_case with PascalCase.
     # pylint: disable=invalid-name
 
-    def __init__(self, stream: io.TextIOBase, *, indent: str):
+    def __init__(self, stream: io.TextIOBase, *, indent: str, chain_else_if: bool = False):
         """
         Args:
             stream (io.TextIOBase): the stream that the output will be written to.
             indent (str): the string to use as a single indentation level.
+            chain_else_if (bool): If ``True``, then constructs of the form::
+
+                    if (x == 0) {
+                        // ...
+                    } else {
+                        if (x == 1) {
+                            // ...
+                        } else {
+                            // ...
+                        }
+                    }
+
+                will be collapsed into the equivalent but flatter::
+
+                    if (x == 0) {
+                        // ...
+                    } else if (x == 1) {
+                        // ...
+                    } else {
+                        // ...
+                    }
+
+                This collapsed form may have less support on backends, as it can be grammatically
+                ambiguous when ``else`` blocks are nested.
```

Comment:
Will this fall back to the unambiguous unchained output when there are nested `else` blocks?

Reply:
It will emit the chained form iff the `else` block contains a single `IfElseOp`.  In those cases, it's unambiguous.  The dangling-else problem is if there's an `else` inside an `if` block (`if (c1): if (c2): h $0; else: x $0;` - what does the `else` bind?), and that's not eligible for being chained here; `if` is always succeeded by `{`, which always disambiguates.

There's `test_chain_else_if` which shows the output of this.